### PR TITLE
Fix blocks number calculation for Flat PA

### PR DESCRIPTION
**Pull request by iboiko-habana**
Fix blocks number calculation for Flat PA via adding empty table_block (https://github.com/HabanaAI/vllm-fork/issues/158)

### CODE REVIEW

File: vllm/worker/habana_model_runner.py [hunk 1]
```diff
@@ -918,7 +923,11 @@ def _prepare_decode(
...
 
                 block_table = seq_group_metadata.block_tables[seq_id]
-                block_number = block_table[position // self.block_size]
+                if len(block_table) == 0:
+                    block_number = 0
+                    block_table = []
+                else:
+                    block_number = block_table[position // self.block_size]
                 if block_number == _PAD_BLOCK_ID:
                     slot = next(dummy_slots)
```
Comment (madamczykhabana) [inline vllm/worker/habana_model_runner.py:932]: Please combine those two if statements. Something like:

```python
if len(block_table) == 0:
    block_number = _PAD_BLOCK_ID
    block_table = []
    slot = next(dummy_slots)
else:
    block_number = block_table[position // self.block_size]
    block_offset = position % self.block_size
    slot = block_number * self.block_size + block_offset
```

**Review by madamczykhabana — CHANGES_REQUESTED** [review]

**Review by madamczykhabana — APPROVED** [review]
LGTM

**Comment by hlin99** [comment]
seq_group_metadata_list.extend(
            self.create_dummy_seq_group_metadata(0, 0, is_prompt)
            for _ in range(batch_size_padding))

this piece of code introduces metadata certation in loop, and observe 10% perf drop. is this code change intentional?

**Comment by iboiko-habana** [comment]
@hlin99 Thanks. Please re-check perf with https://github.com/HabanaAI/vllm-fork/pull/301

**Comment by hlin99** [comment]
> @hlin99 Thanks. Please re-check perf with #301

unfortunately, the perf is not recovered, and perf data looks identical before & after the patch.

i did an experiment as below. the perf could be recovered even if dummy is created each time, but we add seq_group_metadata_list[0] to the list as before. it seems the perf penalty is due to list.extend function.... need further triage

diff --git a/vllm/worker/habana_model_runner.py b/vllm/worker/habana_model_runner.py
index 0d5df1f3..263d55ad 100644
--- a/vllm/worker/habana_model_runner.py
+++ b/vllm/worker/habana_model_runner.py
@@ -1090,7 +1090,7 @@ class HabanaModelRunnerBase(ModelRunnerBase[TModelInputForHPU]):
         if batch_size_padding > 0:
             dummy_seq_group_metadata = self.create_dummy_seq_group_metadata(
                 0, 0, is_prompt)
-            seq_group_metadata_list.extend(dummy_seq_group_metadata
+            seq_group_metadata_list.extend(seq_group_metadata_list[0]
                                            for _ in range(batch_size_padding))

         prefill_reqs = []

**Pull request merged**